core/src/main/java/org/apache/calcite/rel/rules/PruneEmptyRules.java

```diff
@@ -491,6 +513,29 @@ public interface JoinLeftEmptyRuleConfig extends PruneEmptyRule.Config {
     }
   }
 
+  private static void addNullLiterals(RexBuilder rexBuilder, Values empty,
+      List<RexNode> projectFields, List<String> newColumnNames) {
+    for (int i = 0; i < empty.getRowType().getFieldList().size(); ++i) {
+      RelDataTypeField relDataTypeField = empty.getRowType().getFieldList().get(i);
+      RexNode nullLiteral = rexBuilder.makeNullLiteral(relDataTypeField.getType());
+      projectFields.add(nullLiteral);
+      newColumnNames.add(empty.getRowType().getFieldList().get(i).getName());
+    }
+  }
+
+  public static void copyProjects(RexBuilder rexBuilder, RelDataType inRowType,
+      RelDataType castRowType, int castRowTypeOffset,
+      List<RexNode> outProjects, List<String> outProjectNames) {
+    for (int i = 0; i < inRowType.getFieldCount(); ++i) {
+      RelDataTypeField relDataTypeField = inRowType.getFieldList().get(i);
+      RexInputRef inputRef = rexBuilder.makeInputRef(relDataTypeField.getType(), i);
+      RexNode cast = rexBuilder.makeCast(
+          castRowType.getFieldList().get(castRowTypeOffset + i).getType(), inputRef);
+      outProjects.add(cast);
+      outProjectNames.add(relDataTypeField.getName());
+    }
+  }
+
```

Comment:
As highlighted in the corresponding Hive PR, since those two methods are used by both rules, it would be better to move them either before them, or after them, rather than having them in between the rules as of now.

Reply:
Renamed, refactored the methods and also moved after them.